code/network/lib/Loot/Network/Class.hs

```diff
@@ -0,0 +1,176 @@
+  all full nodes are servers).
+
+Server API should:
+* Respond to users' requests asynchronously.
+* Responses should be done in separate "listener" threads, each typically
+  blocking until some request comes and then replying in a free manner until
+  it decides to block again.
+* Publish any (k,v) data, such as transactions, headers or blocks.
+* Send heartbeats to clients.
+Not required:
+* Maintaining list of clients and replying first (without request), that's
+  what "publishing" is for.
+
+-}
+
+module Loot.Network.Class
+    ( Content
+    , Subscription (..)
+    , MsgType (..)
+    , BiTQueue (..)
+
+    , CliRecvMsg(..)
+    , ClientId
+    , ClientEnv
+    , UpdatePeersReq (..)
+    , uprAdd
+    , uprDel
+    , NetworkingCli (..)
+
+    , ListenerId
+    , ListenerEnv
+    , ServSendMsg (..)
+    , NetworkingServ (..)
+    ) where
+
+import Control.Concurrent.STM.TQueue (TQueue)
+import Control.Lens (makeLenses)
+import Data.Default (Default (..))
+
+----------------------------------------------------------------------------
+-- Common
+----------------------------------------------------------------------------
+
+-- | We send and receive lists of strict bytestring. ZMQ
+-- implementation operates in frames, so user can use this extra
+-- functionlaity (for example, attach message numbers in the fisrt
+-- frame and data in the second one). If it is not needed, your
+-- protocol design can state that one or zero BSs are sent.
+type Content = [ByteString]
+
+-- | Bytestring describing the type of the subscription. The key.
+newtype Subscription = Subscription { unSubscription :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Message type is characterized as a bytestring.
+newtype MsgType = MsgType { unMsgType :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Two-ended queue, first parameter is messages user receives,
+-- second -- messages user sends.
+data BiTQueue r s = BiTQueue
+    { bReceiveQ :: TQueue r
+      -- ^ Queue to receive messages from.
+    , bSendQ    :: TQueue s
+      -- ^ Queue to send messages to.
+    }
+
+----------------------------------------------------------------------------
+-- Client
+----------------------------------------------------------------------------
+
+-- | Request to update peers
+data UpdatePeersReq nId = UpdatePeersReq
+    { _uprAdd :: Set nId -- ^ Peers to add/connect.
+    , _uprDel :: Set nId -- ^ Peers to delete/disconnect.
+    } deriving (Show, Generic)
+
+makeLenses ''UpdatePeersReq
+
+instance (Ord s) => Default (UpdatePeersReq s) where
+    def = UpdatePeersReq mempty mempty
+
+-- | Either a response from some server or a subscription update (key
+-- and data).
+data CliRecvMsg
+    = Response MsgType Content
+    | Update Subscription Content
+    deriving (Eq,Ord,Show)
+
+-- | Client worker environment.
+type ClientEnv t = BiTQueue (NodeId t, CliRecvMsg) (Maybe (NodeId t), (MsgType, Content))
```

Comment:
I share your discontent, but I don't think I will resolve this tradeoff by introducing new datatypes.
* All types are different, so it's not possible to mix something up.
* Both receiving and sending packets of client and listener have message types, so I should use field prefixes (which is something I usually try to avoid)
* It's pretty convenient to pattern-match on tuples.

I will add comments though.

Reply:
Ok.